runtime/lua/uri.lua

```diff
@@ -0,0 +1,123 @@
+--- TODO: This is implemented only for files now.
+-- https://tools.ietf.org/html/rfc3986
+-- https://tools.ietf.org/html/rfc2732
+-- https://tools.ietf.org/html/rfc2396
+
+local URI = {}
+
+URI.__index = URI
+URI.new = function(scheme, authority, path, query, fragment)
+  local obj = setmetatable({
+      scheme = scheme,
+      authority = authority,
+      path = path,
+      query = query,
+      fragment = fragment,
+    }, URI)
+  return obj
+end
+
+URI.tostring = function(self)
+  local res = ''
+  local scheme = self.scheme
+  local authority = self.authority
+  local path = self.path
+
+  res = res..scheme..':'
+
+  if authority or scheme == 'file'then
+    if URI.is_windows_filepath(path) then
+      res = res..'///'
+    else
+      res = res..'//'
+    end
+  end
+
+  return res..path
+end
+
+URI.from_filepath = function(path)
```

Comment:
should be called `vim.uri_from_fname` and live in vim.lua. We don't need a separate URI module.

Reply:
@justinmk Where do you mean the `vim.lua` module?

Reply:
Do you mean creating new module 'runtime/lua/vim/lua.lua`?

Reply:
I mean `src/nvim/lua/vim.lua`, but `runtime/lua/vim/shared.lua` is fine as well.

Reply:
I see. I'll try to move all lua modules to proper directory.
Sorry for my unreading vim.lua comments.

Reply:
> I'll try to move all lua modules to proper directory.

Just to be clear:

- LSP-related things can live under `lua/vim/lsp/*` (for now--though I think we don't need ~10 different files in there)
    - `lua/vim/lsp/*` shouldn't have subdirectories (submodules), that is too much hierarchy. So e.g. `lua/vim/lsp/config/*`  should be moved "up".
- `lua/vim/nvim/` doesn't make sense, we already have `lua/vim/` for "core" functionality, thus a separate `nvim` module is confusing. The `vim` module is "the Nvim module".
- My suggestion for URI is to move the few useful functions to `vim.lua`, since they are (or will be...) generally useful.

Reply:
@justinmk 
I understand your intention.

BTW, I don't mean to offend you but I want to hear about your policy.

Why don't you like to have many submodules?
I usually write code in the ruby ​​language, but I think ruby ​​has a tendency to divide it into modules more finely for each role.
I like this policy because I can immediately imagine what the module is for if I look at the file name.

For example, If the methods for uri is in `lua/vim.lua` or `lua/vim/shared.lua`, 
I can't know easily that there is a utility method for uri.
I prefer to `lua/vim/uri.lua` in this point.

I'm sorry for taking your time by asking questions about basic things.

Reply:
People waste time deciding which modules should have which functions, and in the end it doesn't help anything. Everything in `:help eval.txt` is in "one module", and it wouldn't help anything if they were spread around in 20 different modules. This applies to pretty much every project.

Ruby people tend to be obsessed with this because that's what DDH did in Rails.

Reply:
OK, I understand. I am happy to hear your thoughts :)
And I'll write codes with above policy.
Thanks for your time :)

Reply:
I restructured modules.